toolchain/check/eval.cpp

```diff
@@ -0,0 +1,193 @@
+    case SemIR::StructValue::Kind:
+      return TryRebuildIfConstOperands(context, inst,
+                                       &SemIR::StructValue::elements_id);
+    case SemIR::Temporary::Kind:
+      return TryRebuildIfConstOperands(context, inst,
+                                       &SemIR::Temporary::init_id);
+    case SemIR::TupleValue::Kind:
+      return TryRebuildIfConstOperands(context, inst,
+                                       &SemIR::TupleValue::elements_id);
+
+    // These cases are constants already.
+    case SemIR::Builtin::Kind:
+    case SemIR::ClassType::Kind:
+    case SemIR::ConstType::Kind:
+    case SemIR::PointerType::Kind:
+    case SemIR::StructType::Kind:
+    case SemIR::TupleType::Kind:
+    case SemIR::UnboundElementType::Kind:
+      return inst_id;
+
+    case SemIR::BaseDecl::Kind:
+    case SemIR::FieldDecl::Kind:
+    case SemIR::FunctionDecl::Kind:
+      // TODO: Consider adding a corresponding `Value` inst.
+      return inst_id;
+
+    case SemIR::BoolLiteral::Kind:
+    case SemIR::IntLiteral::Kind:
+    case SemIR::RealLiteral::Kind:
+    case SemIR::StringLiteral::Kind:
+      // Promote literals to the constant block.
+      return context.AddConstantInst(inst);
+
+    // TODO: These need special handling.
+    case SemIR::StructInit::Kind:
+    case SemIR::ArrayInit::Kind:
+    case SemIR::BindValue::Kind:
+    case SemIR::Call::Kind:
+    case SemIR::ClassElementAccess::Kind:
+    case SemIR::ClassInit::Kind:
+    case SemIR::Converted::Kind:
+    case SemIR::CrossRef::Kind:
+    case SemIR::Deref::Kind:
+    case SemIR::InitializeFrom::Kind:
+    case SemIR::SpliceBlock::Kind:
+    case SemIR::StructAccess::Kind:
+    case SemIR::TemporaryStorage::Kind:
+    case SemIR::TupleAccess::Kind:
+    case SemIR::TupleIndex::Kind:
+    case SemIR::TupleInit::Kind:
+    case SemIR::ValueAsRef::Kind:
+    case SemIR::ValueOfInitializer::Kind:
+      break;
+
+    case SemIR::BindName::Kind:
+    case SemIR::BindSymbolicName::Kind: {
+      // TODO: Should we really be looking through runtime and symbolic `let`
+      // bindings?
+      return UnwrapIfConstant(context, inst.As<SemIR::AnyBindName>().value_id);
+    }
+
+    case SemIR::NameRef::Kind: {
+      return UnwrapIfConstant(context, inst.As<SemIR::NameRef>().value_id);
+    }
+
+    case SemIR::UnaryOperatorNot::Kind: {
+      auto const_id = UnwrapIfConstant(
+          context, inst.As<SemIR::UnaryOperatorNot>().operand_id);
+      if (!const_id.is_valid()) {
+        break;
+      }
+      auto value = context.insts().TryGetAs<SemIR::BoolLiteral>(const_id);
+      if (!value) {
+        // TODO: Can we CHECK this instead?
+        break;
+      }
+      value->value =
+          (value->value == SemIR::BoolValue::False ? SemIR::BoolValue::True
+                                                   : SemIR::BoolValue::False);
+      return context.AddConstantInst(*value);
+    }
```

Comment:
Added tests for `not` specifically to `check/testdata/operators/unary_op.carbon`. I also did a pass over the other cases and made sure there was testing for the ones that are reachable, but some of these aren't currently reachable because (for example) we don't have any constant-phase reference or initializing expressions yet.

Added a TODO.